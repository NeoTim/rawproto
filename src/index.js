import { Reader } from 'protobufjs'

/**
 * Turn a protobuf into a data-object
 *
 * @param      {Buffer}   buffer  The proto in a binary buffer
 * @return     {object[]}         Info about the protobuf
 */
export const getData = buffer => {
  const reader = Reader.create(buffer)
  const out = []
  while (reader.pos < reader.len) {
    const tag = reader.uint64()
    const id = tag >>> 3
    const wireType = tag & 7
    switch (wireType) {
      case 0: // int32, int64, uint32, bool, enum, etc
        out.push({[id]: reader.uint32()})
        break
      case 1: // fixed64, sfixed64, double
        out.push({[id]: reader.fixed64()})
        break
      case 2: // string, bytes, sub-message
        const bytes = reader.bytes()
        if (bytes[0] === 8) { // this isn't the right way to do this at all, I'm sure
          out.push({[id]: getData(bytes)})
        } else {
          out.push({[id]: bytes.toString()})
        }
        break
      // IGNORE start_group
      // IGNORE end_group
      case 5: // fixed32, sfixed32, float
        out.push({[id]: reader.float()})
        break
      default: reader.skipType(wireType)
    }
  }
  return out
}
