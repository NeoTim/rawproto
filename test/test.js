/* eslint-env jest */
// import { readFileSync, writeFileSync } from 'fs'
import { join } from 'path'
// import { hexy } from 'hexy'
import { getData, getProto, analyzeData } from '../src/index'
import protobuf from 'protobufjs'

let proto

beforeAll(function (done) {
  protobuf.load(join(__dirname, 'demo.proto')).then(p => { proto = p; done() })
})

describe('raw parsing', () => {
  let encoded

  beforeAll(() => {
    const Test = proto.lookupType('Test')
    const msg = Test.create({
      nums: [1, 2, 3, 4, 5],
      num: 1,
      str: 'hello',
      children: [
        {num: 1, str: 'cool', children: [{ num: 1 }]},
        {num: 2, str: 'awesome', children: [{ num: 2 }]},
        {num: 3, str: 'neat', children: [{ num: 3 }]}
      ]
    })
    encoded = Test.encode(msg).finish()
    // console.log('Protobuf:\n', hexy(encoded))
  })

  describe('getData', () => {
    it('parses raw protobuf structure', () => {
      // console.log('Structure:\n', JSON.stringify(getData(encoded), null, 2), '\n')
      expect(getData(encoded)).toMatchSnapshot()
    })
  })

  describe('getProto', () => {
    it('generates example proto definition from raw proto', () => {
      // console.log('Proto:\n', getProto(encoded))
      expect(getProto(encoded)).toMatchSnapshot()
    })
  })
})

// describe('raw proto analysis', function () {
//   it('guesses from set of possible proto definitions', function () {
//     // give it the original .proto but don't tell it which message it's decoding
//     // console.dir(analyzeData(encoded, 'test/demo.proto'), done)
//     // expect(geussData(encoded, readFileSync('demo.proto'))).toMatchObject({
//     //   possibilities: [{
//     //     confidence: 1,
//     //     data:
//     //   }]
//     // })
//   })
// })
