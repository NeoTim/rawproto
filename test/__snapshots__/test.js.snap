// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`raw parsing getData parses raw protobuf structure 1`] = `
Array [
  Object {
    "1": Object {
      "data": Array [
        1,
        2,
        3,
        4,
        5,
      ],
      "type": "Buffer",
    },
  },
  Object {
    "2": 1,
  },
  Object {
    "3": "hello",
  },
  Object {
    "4": Array [
      Object {
        "1": 1,
      },
      Object {
        "2": "cool",
      },
      Object {
        "3": Array [
          Object {
            "1": 1,
          },
        ],
      },
    ],
  },
  Object {
    "4": Array [
      Object {
        "1": 2,
      },
      Object {
        "2": "awesome",
      },
      Object {
        "3": Array [
          Object {
            "1": 2,
          },
        ],
      },
    ],
  },
  Object {
    "4": Array [
      Object {
        "1": 3,
      },
      Object {
        "2": "neat",
      },
      Object {
        "3": Array [
          Object {
            "1": 3,
          },
        ],
      },
    ],
  },
]
`;

exports[`raw parsing getProto generates example proto definition from raw proto 1`] = `
"syntax = \\"proto3\\";

MessageRoot {
  string field1 = 1; // could be a repeated-value, string, buffer, or malformed sub-message
  int32 field2 = 2; // could be a int32, int64, uint32, bool, enum, etc, or even a float of some kind
  string field3 = 3; // could be a repeated-value, string, buffer, or malformed sub-message

  Message4 {
    int32 field1 = 1; // could be a int32, int64, uint32, bool, enum, etc, or even a float of some kind
    string field2 = 2; // could be a repeated-value, string, buffer, or malformed sub-message

    Message3 {
      int32 field1 = 1; // could be a int32, int64, uint32, bool, enum, etc, or even a float of some kind
    }
    
    Message3 subMessage3 = 3;
  }
  
  repeated Message4 subMessage4 = 4;
}"
`;
